### ISSUE TITLE
Connection Manager Valid Transition Order Failure

### ISSUE THREAD

**Issue by bolt12** [input-output-hk/ouroboros-network#3498]
Failure first identified in `bolt12/p2p-master-diffusion-sim` branch

```
    connection_manager_valid_transition_order:                 FAIL (486.22s)
      *** Failed! Falsified (after 79 tests and 125 shrinks):
      0
      ArbDataFlow Duplex
      Script (Script (AbsBearerInfo {abiConnectionDelay = NormalDelay, abiInboundAttenuation = SpeedAttenuation NormalSpeed (Time 46.875s) 716.5s, abiOutboundAttenuation = ErrorInterval FastSpeed (Time 30.166666666666s) 437s, abiInboundWriteFailure = Nothing, abiOutboundWriteFailure = Just 46, abiSDUSize = SmallSDU} :| [AbsBearerInfo {abiConnectionDelay = NormalDelay, abiInboundAttenuation = NoAttenuation NormalSpeed, abiOutboundAttenuation = NoAttenuation NormalSpeed, abiInboundWriteFailure = Nothing, abiOutboundWriteFailure = Nothing, abiSDUSize = NormalSDU}]) :| [])
      MultiNodeScript [StartServer 0s (TestAddr {unTestAddr = TestAddress 47}) 0,StartClient 0s (TestAddr {unTestAddr = TestAddress 61}),InboundConnection 0s (TestAddr {unTestAddr = TestAddress 61}),InboundMiniprotocols 0.1366122644s (TestAddr {unTestAddr = TestAddress 61}) (Bundle {withHot = WithHot [], withWarm = WithWarm [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0], withEstablished = WithEstablished [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]}),StartClient 0.11s (TestAddr {unTestAddr = TestAddress 42}),CloseInboundConnection 0.495s (TestAddr {unTestAddr = TestAddress 61}),StartClient 0.28s (TestAddr {unTestAddr = TestAddress 9}),OutboundConnection 0s (TestAddr {unTestAddr = TestAddress 47}),InboundConnection 0s (TestAddr {unTestAddr = TestAddress 47}),OutboundMiniprotocols 0s (TestAddr {unTestAddr = TestAddress 47}) (Bundle {withHot = WithHot [], withWarm = WithWarm [], withEstablished = WithEstablished []}),InboundConnection 0s (TestAddr {unTestAddr = TestAddress 42})]
         0s: Start server TestAddr {unTestAddr = TestAddress 47} with accInit=0
         0s: Start client TestAddr {unTestAddr = TestAddress 61}
         0s: Connection from TestAddr {unTestAddr = TestAddress 61}
      0.1366122644s: Miniprotocols from TestAddr {unTestAddr = TestAddress 61}: hot:[] warm:[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0] est:[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
      0.2466122644s: Start client TestAddr {unTestAddr = TestAddress 42}
      0.7416122644s: Close connection from TestAddr {unTestAddr = TestAddress 61}
      1.0216122644s: Start client TestAddr {unTestAddr = TestAddress 9}
      1.0216122644s: Connecting to TestAddr {unTestAddr = TestAddress 47}
      1.0216122644s: Connection from TestAddr {unTestAddr = TestAddress 47}
      1.0216122644s: Miniprotocols to TestAddr {unTestAddr = TestAddress 47}: hot:[] warm:[] est:[]
      1.0216122644s: Connection from TestAddr {unTestAddr = TestAddress 42}
      TransitionTrace @(TestAddress 61) (UnknownConnectionSt → UnnegotiatedSt Inbound)
      TransitionTrace @(TestAddress 61) (UnnegotiatedSt Inbound → InboundIdleSt Unidirectional)
      TransitionTrace @(TestAddress 61) (InboundIdleSt Unidirectional → InboundSt Unidirectional)
      TransitionTrace @(TestAddress 61) (InboundSt Unidirectional → InboundIdleSt Unidirectional)
      TransitionTrace @(TestAddress 61) (InboundIdleSt Unidirectional → InboundSt Unidirectional)
      TransitionTrace @(TestAddress 61) (InboundSt Unidirectional → TerminatedSt)
      TransitionTrace @(TestAddress 61) (TerminatedSt → UnknownConnectionSt)
      TransitionTrace @(TestAddress 47) (UnknownConnectionSt → ReservedOutboundSt)
      TransitionTrace @(TestAddress 47) (ReservedOutboundSt → UnnegotiatedSt Inbound)
      TransitionTrace @(TestAddress 42) (UnknownConnectionSt → UnnegotiatedSt Inbound)
      TransitionTrace @(TestAddress 42) (UnnegotiatedSt Inbound → TerminatedSt)
      TransitionTrace @(TestAddress 42) (TerminatedSt → UnknownConnectionSt)
      TransitionTrace @(TestAddress 47) (UnnegotiatedSt Inbound → TerminatedSt)
      TransitionTrace @(TestAddress 47) (TerminatedSt → UnknownConnectionSt)
      TransitionTrace @(TestAddress 47) (TerminatingSt → TerminatedSt)
      MainReturn (Time 3601.0216122644s) () []
      
      Unexpected last transition: TerminatingSt → TerminatedSt
      Use --quickcheck-replay=610691 to reproduce.
```

**Comment by coot** [comment]
This is fixed in 7a3d65c59f1a437e1ee3088bb0d10845c0c0bd96.

**Comment by coot** [comment]
Just for historical reasons, since it's a very good example of a bug discovered with `IOSim`: the problem was that we updated a connection state where it shouldn't be: there are two threads that can move the connection state: either a thread which calls one of connection manager methods or the clean-up handler of connection handler thread.
